cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java

```diff
@@ -466,80 +469,87 @@ ReconciliationState createReconciliationState(Reconciliation reconciliation, Kaf
+            return waitForQuiescence(namespace, kafkaSsName).compose(quiescence -> {
+                if (!quiescence) {
+                    return Future.succeededFuture(this);
+                } else {
+                    return kafkaSetOperations.getAsync(namespace, kafkaSsName).compose(ss -> {
+                        if (ss == null) {
+                            return Future.succeededFuture(this);
+                        }
+                        log.debug("Does SS {} need to be upgraded?", ss.getMetadata().getName());
+                        Future<?> result;
+                        // Get the current version of the cluster
+                        // Strimzi 0.8.x and 0.9.0 didn't set the annotation, so if it's absent we know it must be 2.0.0
+                        KafkaVersion currentVersion = versions.version(Annotations.annotations(ss).getOrDefault(ANNO_STRIMZI_IO_KAFKA_VERSION, "2.0.0"));
+                        log.debug("SS {} has current version {}", ss.getMetadata().getName(), currentVersion);
+                        String fromVersionAnno = Annotations.annotations(ss).get(ANNO_STRIMZI_IO_FROM_VERSION);
+                        KafkaVersion fromVersion;
+                        if (fromVersionAnno != null) { // We're mid-upgrade
+                            fromVersion = versions.version(fromVersionAnno);
                         } else {
-                            if (currentVersion.equals(fromVersion)) {
-                                f = f.compose(ignored -> kafkaDowngradePhase1(ss, upgrade));
-                            }
-                            result = f.compose(ignored -> kafkaDowngradePhase2(ss, upgrade, image));
+                            fromVersion = currentVersion;
+                        }
+                        log.debug("SS {} is from version {}", ss.getMetadata().getName(), fromVersion);
+                        String toVersionAnno = Annotations.annotations(ss).get(ANNO_STRIMZI_IO_TO_VERSION);
+                        KafkaVersion toVersion;
+                        if (toVersionAnno != null) { // We're mid-upgrade
+                            toVersion = versions.version(toVersionAnno);
+                        } else {
+                            toVersion = versions.version(kafkaAssembly.getSpec().getKafka().getVersion());
                         }
+                        log.debug("SS {} is to version {}", ss.getMetadata().getName(), toVersion);
+                        KafkaUpgrade upgrade = new KafkaUpgrade(fromVersion, toVersion);
+                        log.debug("Kafka upgrade {}", upgrade);
+                        if (upgrade.isNoop()) {
+                            log.debug("Kafka.spec.kafka.version unchanged");
+                            result = Future.succeededFuture();
+                        } else {
+                            String image = versions.kafkaImage(kafkaAssembly.getSpec().getKafka().getImage(), toVersion.version());
```

Comment:
What I expected as the solution was to keep the old `waitForQuiescence` here. Here we already know that we need to do the upgrade so we can ensure the brokers are ready for it.

Reply:
The problem that I saw was a no need to roll for a pod which was crashing and the readiness blocked indefinitely, can't it happen again?

Reply:
Ic oult happen again ... but only in situation when the upgrade is actually happening ... i.e. it will happen only when needed. Where as the original implementation blocks it even when the upgrade is actually completely irrelevant.

Reply:
So we don't solve the problem when upgrade should happen

Reply:
Maybe we do not need to solve it then. And also the probability that this happens all together is fairly low.